File: Linear_Control_Course/Chapter25/Lesson1/Chapter25_Lesson1.java
public class CascadeController {
    private double k1_p; // inner gain (velocity)
    private double k2_p; // outer gain (position)
    private double dt;

    private double v;    // velocity state
    private double x;    // position state

    public CascadeController(double k1_p, double k2_p, double dt) {
        this.k1_p = k1_p;
        this.k2_p = k2_p;
        this.dt = dt;
        this.v = 0.0;
        this.x = 0.0;
    }

    // One simulation step given desired position rPos
    public void step(double rPos) {
        double ePos = rPos - x;
        double vRef = k2_p * ePos;

        double eVel = vRef - v;
        double u = k1_p * eVel;

        // Simple plant model: dv/dt = -alpha * v + beta * u
        double alpha = 50.0;
        double beta = 50.0;
        double dv = -alpha * v + beta * u;

        v += dv * dt;
        x += v * dt;
    }

    public double getPosition() {
        return x;
    }

    public double getVelocity() {
        return v;
    }

    public static void main(String[] args) {
        CascadeController ctrl = new CascadeController(15.0, 4.0, 0.001);
        double rPos = 1.0;

        for (int k = 0; k < 5000; ++k) {
            ctrl.step(rPos);
            if (k % 1000 == 0) {
                System.out.println("t = " + (k * 0.001)
                        + "  x = " + ctrl.getPosition()
                        + "  v = " + ctrl.getVelocity());
            }
        }
    }
}

File: Tutorials/Codes/Linear_Control_Course/Chapter25/Lesson1/Chapter25_Lesson1.cpp
#include <iostream>

int main() {
    const double dt = 0.001;      // sampling time [s]
    const int N = 5000;           // simulation steps

    // Controller gains
    const double k1_p = 20.0;     // inner (velocity) P gain
    const double k2_p = 5.0;      // outer (position) P gain

    // Simple first-order plant model for velocity
    double v = 0.0;               // velocity state
    const double a = -50.0;       // pole coefficient
    const double b = 50.0;        // input gain

    double x = 0.0;               // position (integral of velocity)

    double r_pos = 1.0;           // desired position

    for (int k = 0; k < N; ++k) {
        // Outer loop: position control
        double e_pos = r_pos - x;
        double v_ref = k2_p * e_pos;     // desired velocity

        // Inner loop: velocity control
        double e_vel = v_ref - v;
        double u = k1_p * e_vel;        // actuator command (e.g., torque)

        // Plant update (Euler discretization)
        double dv = a * v + b * u;
        v += dv * dt;
        x += v * dt;

        if (k % 1000 == 0) {
            std::cout << "t = " << k * dt
                      << "  x = " << x
                      << "  v = " << v
                      << std::endl;
        }
    }
    return 0;
}
